File: src/test/java/ca/yyx/hu/aap/AapReadTest.java
package ca.yyx.hu.aap;

import android.support.annotation.NonNull;

import org.junit.Test;

import java.nio.ByteBuffer;

import ca.yyx.hu.aap.protocol.Channel;
import ca.yyx.hu.aap.protocol.MsgType;
import ca.yyx.hu.connection.AccessoryConnection;
import ca.yyx.hu.utils.AppLog;
import ca.yyx.hu.utils.Utils;

import static org.junit.Assert.assertEquals;

/**
 * @author algavris
 * @date 13/02/2017.
 */

public class AapReadTest {

    class FakeConnection implements AccessoryConnection {
        byte[] data = new byte[0];

        @Override
        public boolean isSingleMessage() {
            return false;
        }

        @Override
        public int send(byte[] buf, int length, int timeout) {
            return 0;
        }

        @Override
        public int recv(byte[] buf, int length, int timeout) {
            System.arraycopy(this.data, 0, buf, 0, this.data.length);
            return this.data.length;
        }

        @Override
        public boolean isConnected() {
            return false;
        }

        @Override
        public void connect(@NonNull Listener listener) {

        }

        @Override
        public void disconnect() {

        }
    }

    class FakeHandler implements AapMessageHandler {
        AapMessage[] msg = new AapMessage[10];
        int idx = 0;
        @Override
        public void handle(AapMessage message) throws HandleException {
            this.msg[this.idx] = message;
            this.idx++;
        }
    }

    class FakeSsl implements AapSsl {

        @Override
        public int prepare() {
            return 0;
        }

        @Override
        public void handshake() {

        }

        @Override
        public ByteArray bioRead() {
            return null;
        }

        @Override
        public int bioWrite(int start, int length, byte[] buffer) {
            return 0;
        }

        @Override
        public ByteArray decrypt(int start, int length, byte[] buffer) {
            byte[] result = new byte[length];
            System.arraycopy(buffer, start, result, 0, length);
            return new ByteArray(result, length);
        }

        @Override
        public ByteArray encrypt(int offset, int length, byte[] buffer) {
            return null;
        }
    }

    @Test
    public void readMultipleMessages() {
        AppLog.LOGGER = new AppLog.Logger.StdOut();
        FakeConnection conn = new FakeConnection();
        FakeSsl ssl = new FakeSsl();
        FakeHandler handler = new FakeHandler();
        AapReadMultipleMessages poll = new AapReadMultipleMessages(conn, ssl, handler);

        conn.data = new byte[] {
            Channel.ID_CTR, 0x8, 0, 6, 0, MsgType.Control.PINGRESPONSE, 'M', 'S', 'G', '1',
            Channel.ID_VID, 0x9, 0, 7, 0, 0, 0, 1, 0, MsgType.Control.MEDIADATA, 'V', 'I', 'D', 'E', 'O',
            Channel.ID_CTR, 0x8, 0, 6, 0, MsgType.Control.PINGRESPONSE, 'M',
        };

        int res = poll.read();
        assertEquals(res, 0);
        assertEquals(handler.idx, 2);
        assertEquals(handler.msg[0], new AapMessage(Channel.ID_CTR, (byte) 0x8, MsgType.Control.PINGRESPONSE, 2, 6, new byte[] { 0, MsgType.Control.PINGRESPONSE, 'M', 'S', 'G', '1' }));
        assertEquals(handler.msg[1], new AapMessage(Channel.ID_VID, (byte) 0x9, MsgType.Control.MEDIADATA, 2, 7, new byte[] { 0, MsgType.Control.MEDIADATA, 'V', 'I', 'D', 'E', 'O' }));

        conn.data = new byte[] {
                'S', 'G', '2',
                Channel.ID_CTR, 0x8,
        };

        res = poll.read();
        assertEquals(res, 0);
        assertEquals(handler.idx, 3);
        assertEquals(handler.msg[2], new AapMessage(Channel.ID_CTR, (byte) 0x8, MsgType.Control.PINGRESPONSE, 2, 6, new byte[] { 0, MsgType.Control.PINGRESPONSE, 'M', 'S', 'G', '2' }));

        conn.data = new byte[] {
                 0, 6, 0, MsgType.Control.PINGRESPONSE, 'M', 'S', 'G', '3',
        };

        res = poll.read();
        assertEquals(res, 0);
        assertEquals(handler.idx, 4);
        assertEquals(handler.msg[3], new AapMessage(Channel.ID_CTR, (byte) 0x8, MsgType.Control.PINGRESPONSE, 2, 6, new byte[] { 0, MsgType.Control.PINGRESPONSE, 'M', 'S', 'G', '3' }));
    }
}

File: src/test/java/ca/yyx/hu/aap/AapReadTest.kt
package ca.yyx.hu.aap

import org.junit.Test

import ca.yyx.hu.aap.protocol.Channel
import ca.yyx.hu.aap.protocol.MsgType
import ca.yyx.hu.connection.AccessoryConnection
import ca.yyx.hu.utils.AppLog

import org.junit.Assert.assertEquals

/**
 * @author algavris
 * *
 * @date 13/02/2017.
 */

class AapReadTest {

    internal inner class FakeConnection : AccessoryConnection {
        var data = ByteArray(0)

        override val isSingleMessage: Boolean
            get() = false

        override fun send(buf: ByteArray, length: Int, timeout: Int): Int {
            return 0
        }

        override fun recv(buf: ByteArray, length: Int, timeout: Int): Int {
            System.arraycopy(this.data, 0, buf, 0, this.data.size)
            return this.data.size
        }

        override val isConnected: Boolean
            get() = false

        override fun connect(listener: AccessoryConnection.Listener) {

        }

        override fun disconnect() {

        }
    }

    internal inner class FakeHandler : AapMessageHandler {
        var msg = arrayOfNulls<AapMessage>(10)
        var idx = 0
        @Throws(AapMessageHandler.HandleException::class)
        override fun handle(message: AapMessage) {
            this.msg[this.idx] = message
            this.idx++
        }
    }

    internal inner class FakeSsl : AapSsl {

        override fun prepare(): Int {
            return 0
        }

        override fun handshake() {

        }

        override fun bioRead(): ByteArrayWithLimit? {
            return null
        }

        override fun bioWrite(start: Int, length: Int, buffer: ByteArray): Int {
            return 0
        }

        override fun decrypt(start: Int, length: Int, buffer: ByteArray): ByteArrayWithLimit? {
            val result = ByteArray(length)
            System.arraycopy(buffer, start, result, 0, length)
            return ByteArrayWithLimit(result, length)
        }

        override fun encrypt(offset: Int, length: Int, buffer: ByteArray): ByteArrayWithLimit? {
            return null
        }
    }

    @Test
    fun readMultipleMessages() {
        AppLog.LOGGER = AppLog.Logger.StdOut()
        val conn = FakeConnection()
        val ssl = FakeSsl()
        val handler = FakeHandler()
        val poll = AapReadMultipleMessages(conn, ssl, handler)

        conn.data = byteArrayOf(Channel.ID_CTR.toByte(), 0x8, 0, 6, 0, MsgType.Control.PINGRESPONSE.toByte(), b('M'), b('S'), b('G'), b('1'), Channel.ID_VID.toByte(), 0x9, 0, 7, 0, 0, 0, 1, 0, MsgType.Control.MEDIADATA.toByte(), b('V'), b('I'), b('D'), b('E'), b('O'), Channel.ID_CTR.toByte(), 0x8, 0, 6, 0, MsgType.Control.PINGRESPONSE.toByte(), b('M'))

        var res = poll.read()
        assertEquals(res.toLong(), 0)
        assertEquals(handler.idx.toLong(), 2)
        assertEquals(handler.msg[0], AapMessage(Channel.ID_CTR, 0x8.toByte(), MsgType.Control.PINGRESPONSE, 2, 6, byteArrayOf(0, MsgType.Control.PINGRESPONSE.toByte(), b('M'), b('S'), b('G'), b('1'))))
        assertEquals(handler.msg[1], AapMessage(Channel.ID_VID, 0x9.toByte(), MsgType.Control.MEDIADATA, 2, 7, byteArrayOf(0, MsgType.Control.MEDIADATA.toByte(), b('V'), b('I'), b('D'), b('E'), b('O'))))

        conn.data = byteArrayOf(b('S'), b('G'), b('2'), Channel.ID_CTR.toByte(), 0x8)

        res = poll.read()
        assertEquals(res.toLong(), 0)
        assertEquals(handler.idx.toLong(), 3)
        assertEquals(handler.msg[2], AapMessage(Channel.ID_CTR, 0x8.toByte(), MsgType.Control.PINGRESPONSE, 2, 6, byteArrayOf(0, MsgType.Control.PINGRESPONSE.toByte(), b('M'), b('S'), b('G'), b('2'))))

        conn.data = byteArrayOf(0, 6, 0, MsgType.Control.PINGRESPONSE.toByte(), b('M'), b('S'), b('G'), b('3'))

        res = poll.read()
        assertEquals(res.toLong(), 0)
        assertEquals(handler.idx.toLong(), 4)
        assertEquals(handler.msg[3], AapMessage(Channel.ID_CTR, 0x8.toByte(), MsgType.Control.PINGRESPONSE, 2, 6, byteArrayOf(0, MsgType.Control.PINGRESPONSE.toByte(), b('M'), b('S'), b('G'), b('3'))))
    }
}

private fun b(ch: Char): Byte { return ch.toByte() }